packages/catlog/src/stdlib/analyses/ecld/atomisations.rs

```diff
@@ -0,0 +1,292 @@
+
+    // height: the total height of the tower
+    // in_arrows_pos_deg: all incoming arrows of (strictly) positive degree
+    // in_arrows_zero_deg: all incoming arrows of zero degree
+    struct TowerConstructor {
+        height: usize,
+        in_arrows_pos_deg: Vec<QualifiedName>,
+        in_arrows_zero_deg: Vec<QualifiedName>,
+    }
+    let mut tower_constructors: HashMap<QualifiedName, TowerConstructor> = HashMap::new();
+
+    // Every tower will be of height at least 1, and will have at the very least
+    // an empty list of positive-degree (resp. zero-degree) incoming arrows
+    for x in model.ob_generators() {
+        tower_constructors.insert(
+            x.clone(),
+            TowerConstructor {
+                height: 1,
+                in_arrows_pos_deg: Vec::new(),
+                in_arrows_zero_deg: Vec::new(),
+            },
+        );
+    }
+
+    // -------------------------------------------------------------------------
+    // 1.   For each base, calculate the maximum degree over all incoming
+    //      arrows. Note that this is most easily done by actually iterating
+    //      over the *morphisms* instead.
+
+    for f in model.mor_generators() {
+        let f_cod = model.get_cod(&f).unwrap();
+        let f_degree = deg_of_mor(&model, &f);
+
+        if f_degree != 0 {
+            let new_degree = std::cmp::max(tower_constructors.get(f_cod).unwrap().height, f_degree);
+            tower_constructors
+                .entry(f_cod.clone())
+                .and_modify(|tower_cons| tower_cons.height = new_degree)
+                // Since we already have the codomain, we can add this to our
+                // hash map of positive-degree incoming arrows
+                .and_modify(|tower_cons| tower_cons.in_arrows_pos_deg.push(f.clone()));
+        } else {
+            tower_constructors
+                .entry(f_cod.clone())
+                .and_modify(|tower_cons| tower_cons.in_arrows_zero_deg.push(f.clone()));
+        }
+    }
+
+    // -------------------------------------------------------------------------
+    // 2.   Iterate over all unchecked bases, starting with (any one of) the
+    //      one(s) with greatest current height. For each one, ensure that all
+    //      the incoming arrows can be lifted so that their target is the
+    //      highest floor of the tower, i.e. that the tower over Y has enough
+    //      floors so that *every* arrow into Y can have the same target Y_max.
+    //      But note that if we lift an arrow of degree n from X to Y then we
+    //      need to know that X_n exists, i.e. we need to (potentially) add
+    //      more floors to the *source* of every arrow whose target is Y.
+
+    // We have yet to build any of the towers so, right now, every base is
+    // "unchecked".
+    let mut unchecked_bases: Vec<QualifiedName> = model.ob_generators().collect::<Vec<_>>().clone();
```

Comment:
Never need to `clone` immediately after `collect`: you're already getting a fresh object.